Display author avatars in the blog

rowanmanning opened the pull request:
This resolves #1758 by adding author images to blog posts, both on the post page and in the list view. It's not overly designed - I went for a simple solution for now that can be expanded on later if the site is refreshed (as suggested in #1787).

This required some foundational work to ensure that:

  * We support multiple authors on a post with different metadata. I also updated all existing posts to ensure that the metadata is in place
  * Moving the author list to an include so that we don't have to repeat a load of Liquid logic in two places on the site

I outlined this work in more detail in [this comment on the original issue](https://github.com/expressjs/expressjs.com/issues/1758#issuecomment-2820523545).

## Screenshots

Overview page with single author:

<img width="896" alt="Screenshot 2025-05-27 at 15 49 14" src="https://github.com/user-attachments/assets/98cdecc1-010f-41cd-ac9b-d6038cf6a98f" />

Overview page with one author with an avatar and one without:

<img width="896" alt="Screenshot 2025-05-27 at 15 49 14" src="https://github.com/user-attachments/assets/695419f8-8fff-47eb-b3ef-ea9994cfc506" />

Overview page with two authors:

<img width="900" alt="Screenshot 2025-05-27 at 15 49 23" src="https://github.com/user-attachments/assets/f8592e9c-eb25-411c-a88a-0848c39c66bb" />

Light mode:

<img width="899" alt="Screenshot 2025-05-27 at 15 49 33" src="https://github.com/user-attachments/assets/0fd15764-4093-469b-96ce-ab4b657a666d" />

Blog post page:

<img width="907" alt="Screenshot 2025-05-27 at 15 49 47" src="https://github.com/user-attachments/assets/3c869ec7-e01d-4a4e-91b3-d1bef463497b" />

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

rowanmanning committed:
feat: allow blogs to have multiple authors

This requires us to update the `author` front-matter, requiring it to be
an array of objects instead of a single string. This paves the way for
us to include further author metadata.

For now we only add "name" which expects a string.

rowanmanning committed:
feat: allow linking to author GitHub profiles

This adds the ability for authors to add their GitHub username to the
blog posts they author. If they provide this then their name will be
linked to their GitHub profile page.

rowanmanning committed:
feat: add GitHub profile images to author lines

This adds a small profile image to author lines that have a GitHub.

rowanmanning committed:
feat: add author URLs to the feeds

wesleytodd commented on css/style.css:
```diff
@@ -1487,9 +1487,22 @@ h2 a {
+  border-radius: 50%;
+  display: inline;
+  vertical-align: middle;
+  width: 1.5em;
```
Could we just push the avatar over a tiny bit to match the spacing on both sides? I have no idea if this is the right value (aka it is probably too much) but figured I would leave it up to y'all to get the spacing right.

```suggestion
  width: 1.5em;
  padding: .25em;
```

rowanmanning replied:
Done 👍 I added a half character of left margin which events things out better. It now looks like this:

![Screenshot 2025-05-28 at 09 32 08](https://github.com/user-attachments/assets/b8867bb7-2aa8-42c6-ac63-8de26c6a69c1)

bjohansebas commented on _includes/feed-entry.xml:
```diff
@@ -3,11 +3,12 @@
+    {% if author.github %}<uri>https://github.com/{{ author.github }}</uri>{% endif %}
   </author>
-  {%- endif %}
+  {%- endfor %}
```
Could you separate the code better? I know we don't have a linter, but it's hard to read the code.

rowanmanning replied:
Done 👍

bjohansebas commented on _posts/2024-10-15-v5-release.md:
```diff
@@ -1,7 +1,10 @@
+authors:
+  - name: Wes Todd
+    github: wesleytodd
+  - name: Express Technical Committee
```
We could use the GitHub username for the TC, so it doesn't end up without a photo.

rowanmanning replied:
I couldn't find a GitHub user for the technical committee specifically, I added [expressjs](https://github.com/expressjs) and it looks like this:

![Screenshot 2025-05-28 at 09 42 06](https://github.com/user-attachments/assets/a2398bfd-db44-47b1-a42b-1b0465ecc288)

Is this what you meant? If not I can revert. One doubt I had is that we're kind of spamming the blog page with lots of the same image and link out to the GitHub org. Maybe this is fine?

rowanmanning committed:
fix: space the authors a little more

The authors looked a little bunched up, a left margin on each of them
reads a little better.

Co-Authored-By: Wes Todd <wesleytodd@users.noreply.github.com>

rowanmanning committed:
fix: improve readability of the feed entry Liquid

Co-Authored-By: Sebastian Beltran <bjohansebas@users.noreply.github.com>

rowanmanning committed:
feat: add author image for the technical committee

Co-Authored-By: Sebastian Beltran <bjohansebas@users.noreply.github.com>

ShubhamOulkar commented on css/style.css:
```diff
@@ -1487,9 +1487,23 @@ h2 a {
+  border-radius: 50%;
+  display: inline;
+  vertical-align: middle;
+  width: 1.5em;
```
```suggestion
  width: 2em;
  height: 2em;
```
32px avatar looks good

rowanmanning replied:
I went with the smaller size because it matched up more with the existing line height so it didn't shift any elements vertically. With `2em` the image no longer aligns with the text, pushing the height of the post metadata to be larger than a regular paragraph. I think the post view looks a little janky with larger images. Happy to go with it if others disagree but would like more opinions.

![Screenshot 2025-05-28 at 14 20 11](https://github.com/user-attachments/assets/5b83b441-9c04-4045-85af-fbab4e806bff)

ShubhamOulkar replied:
> image no longer aligns with the text

most of the websites use not aligned image with the text. text always align at center with respect to the image. [example](https://developer.mozilla.org/en-US/blog/color-models-humans-devices/).

ShubhamOulkar commented on css/style.css:
```diff
@@ -1487,9 +1487,23 @@ h2 a {
+}
+.blog-author-avatar {
+  border-radius: 50%;
+  display: inline;
```
```suggestion
```
Images are `inline-block` by default and `inline` prevents height or aspect-ratio from working properly. Also set equal width and height for a perfect circle. GitHub avatars are perfect squares, so the current border-radius: 50% works fine now, but this ensures it stays consistency in future design updated.

rowanmanning replied:
Done 👍

rowanmanning committed:
fix: set a width and height on author images

This ensures we still display a circle if, for some reason, the image
that we get isn't a perfect square.

Co-Authored-By: Shubham Oulkar <ShubhamOulkar@users.noreply.github.com>

ShubhamOulkar commented on _includes/blog/authors.html:
```diff
@@ -0,0 +1,21 @@
+
+      {% if label and author.github %}
+        <a href="https://github.com/{{author.github}}" rel="author" class="blog-author-link">
+          <img src="https://github.com/{{author.github}}.png?size=64" alt="" class="blog-author-avatar" />
```
```suggestion
          <img src="https://github.com/{{author.github}}.png?size=64" alt="{{label}} avatar" class="blog-author-avatar" />
```

rowanmanning replied:
I struggled to find some good "official" guidance on this, across other websites:

  * MDN includes a generic "Author avatar"
  * GitHub includes a generic "Owner avatar" in some places and the username in others
  * Mastodon includes an empty `alt` attribute

So things aren't consistent. Then I tried using VoiceOver on my Mac to see what the actual result is. With GitHub in a place like this:

![Screenshot 2025-05-28 at 15 41 46](https://github.com/user-attachments/assets/7d0003f9-76c5-441e-b774-374a1572d346)

The screen reader announces `link image @rowanmanning` then I tab to the name and it announces `link @rowanmanning`. The extra information is offering no real value to a screen reader user. I tried your suggestion with my local copy of the site and you get a similar experience:

> "link image Ulises Gascón avatar [tab] link Ulises Gascón"

Again, I think the information is redundant. In our case the avatar is purely visual because it has the user's name/link right next to it. We can't easily include an `alt` text that actually describes the content of the image ([as this blog post suggests](https://nicolas-steenhout.com/alt-text-for-avatars/)) so the (admittedly lazy) but next best thing is to get screen readers to ignore it.

ShubhamOulkar replied:
Thank you for correcting me. You are right, I didn't consider image is children of anchor links.

ShubhamOulkar commented on _includes/blog/authors.html:
```diff
@@ -0,0 +1,21 @@
+      {%- endif -%}
+
+      {% if label and author.github %}
+        <a href="https://github.com/{{author.github}}" rel="author" class="blog-author-link">
```
```suggestion
        <a href="https://github.com/{{author.github}}" rel="author" class="blog-author-link" target="_blank">
```

rowanmanning replied:
I avoided target `_blank` because my understanding is that it's bad accessibility practice – it breaks the [change on request](https://www.w3.org/WAI/WCAG21/Understanding/change-on-request.html) success criterion. From the linked guide:

> This success criterion aims to eliminate potential confusion that may be caused by unexpected changes of context **such as automatic launching of new windows**...

Anecdotally I work with an engineer with peripheral vision loss who has moaned to me about this practice in the past 😄

The [related guidance](https://www.w3.org/TR/WCAG20-TECHS/G201.html) says that, if you're opening in a new window with `_blank` then you should either provide a visual indicator (this will clutter the interface in my opinion) or include text like "opens in a new window". For me the lowest-code solution is just to let users decide where the link gets opened.

ShubhamOulkar replied:
Target `_blank` has specific use cases like visiting author profile. You can verify it from any blog/documentation website.

rowanmanning replied:
Yep I agree it's _fairly_ common, but it's not good accessibility practice as your original comment suggested. I'd be more on the side of making this accessible than copying the use of `_blank` from other sites - this doesn't fit into any of the ["good reasons" here](https://css-tricks.com/use-target_blank/).

Here are a few blogs that don't open the author link in a new tab to illustrate it's not _everyone_ doing this (grabbed from recent reading): [The ESLint blog](https://eslint.org/blog/2025/05/eslint-v9.27.0-released/), [The Docker blog](https://www.docker.com/blog/introducing-docker-hardened-images/) - but it does provide its own author pages.

ShubhamOulkar replied:
Ok agreed

ShubhamOulkar reviewed (changes requested):
Accessibility improvements

ShubhamOulkar reviewed (approved):
LGTM 
Please resolve merge conflict

bjohansebas reviewed (approved):
LGTM

bjohansebas commented:
thanks @rowanmanning!
